### PR TITLE
[feature] python support 

### PR DESCRIPTION
**Pull request by mielnikk**

### CODE REVIEW

File: e2e/test-resources/python-project/.bazelversion [hunk 1]
```diff
@@ -0,0 +1 @@
+5.1.0
```
Comment (abrams27) [inline e2e/test-resources/python-project/.bazelversion:1]: u dont need it - the ci tests will run it with bazel 5 and 6

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspPythonProjectTest.kt [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    @JvmStatic
+    fun main(args: Array<String>) = executeScenario()
+
+    override fun repoName(): String = "python-project"
+
+    override fun scenarioSteps(): List<BazelBspTestScenarioStep> = listOf(workspaceBuildTargets(), dependencySourcesResults())
+
+    private fun workspaceBuildTargets(): BazelBspTestScenarioStep {
+
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspPythonProjectTest.kt:18]:
```suggestion
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspPythonProjectTest.kt [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            listOf("python"),
+            listOf(BuildTargetIdentifier("@pip_deps_numpy//:pkg")),
+            BuildTargetCapabilities(true, false, false, false)
+        )
+        pythonLibBuildTarget.displayName = "$targetPrefix//lib:example_library"
+        pythonLibBuildTarget.baseDirectory = "file://\$WORKSPACE/lib/"
+        pythonLibBuildTarget.data = pythonLibPythonBuildTarget
+        pythonLibBuildTarget.dataKind = BuildTargetDataKind.PYTHON
+
+        val expectedWorkspaceBuildTargetsResult = WorkspaceBuildTargetsResult(listOf(pythonLibBuildTarget))
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspPythonProjectTest.kt:118]: u can just reuse the previous method

File: e2e/test-resources/python-project/example/BUILD [hunk 4]
```diff
@@ -0,0 +1,8 @@
+load("@rules_python//python:defs.bzl", "py_binary")
+
+py_binary(
+    name = "example",
+    srcs = ["example.py"],
+    python_version = "PY3",
+    deps = ["//lib:example_library", "@requests//:srcs"],
```
Comment (abrams27) [inline e2e/test-resources/python-project/example/BUILD:7]: let's put each dep in a separate line

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonModule.kt [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+import org.jetbrains.bsp.bazel.server.sync.languages.LanguageData
+import java.net.URI
+
+data class PythonModule(
+    val interpreter: URI?,
+    val version: String?
+) : LanguageData
+
+
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonModule.kt:11]:
```suggestion
```

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import kotlin.io.path.toPath
+
+class PythonLanguagePlugin(
+    private val bazelPathsResolver: BazelPathsResolver
+) : LanguagePlugin<PythonModule>() {
+
+    override fun resolveModule(targetInfo: TargetInfo): PythonModule? =
+        targetInfo.pythonTargetInfo?.run {
+
+            val interpreterURI = interpreter?.let {
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:26]: can we extract the calculation to a method?

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                interpreterURI,
+                version.takeUnless(String::isNullOrEmpty)
+            )
+
+        }
+
+
+    override fun applyModuleData(moduleData: PythonModule, buildTarget: BuildTarget) {
+        buildTarget.dataKind = BuildTargetDataKind.PYTHON
+        val interpreter = moduleData.interpreter?.let { it.toString() }
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:40]:
```suggestion
        val interpreter = moduleData.interpreter?.toString()
```

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            emptyList(),
+        )
+
+    override fun dependencySources(targetInfo: TargetInfo, dependencyTree: DependencyTree): Set<URI> {
+        return targetInfo.pythonTargetInfo?.run {
+            dependencyTree.transitiveDependenciesWithoutRootTargets(targetInfo.id)
+                .flatMap(::getExternalSources)
+                .map(bazelPathsResolver::resolveUri)
+                .map {
+                    fun findExternal(path: Path?): Path? {
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:56]: same here

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+                    val path = it.toPath()
+                    bazelPathsResolver.resolveUri(findExternal(path) ?: path)
+                }
+                .toSet()
+        }.orEmpty()
+    }
+
+    private fun getExternalSources(targetInfo: TargetInfo): List<FileLocation> {
+        return targetInfo.sourcesList.mapNotNull { it.takeIf { it.isExternal } }
+    }
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:72]:
```suggestion
    private fun getExternalSources(targetInfo: TargetInfo): List<FileLocation> =
        targetInfo.sourcesList.mapNotNull { it.takeIf { it.isExternal } }
```

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+    fun toPythonOptionsItem(module: Module, pythonModule: PythonModule): PythonOptionsItem =
+        PythonOptionsItem(
+            BspMappings.toBspId(module),
+            emptyList(),
+        )
+
+    override fun dependencySources(targetInfo: TargetInfo, dependencyTree: DependencyTree): Set<URI> {
+        return targetInfo.pythonTargetInfo?.run {
+            dependencyTree.transitiveDependenciesWithoutRootTargets(targetInfo.id)
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:52]: and here

**Review by abrams27 — COMMENTED** [review]
looks nice, thanks! 

just some nitpicking

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    fun toPythonOptionsItem(module: Module, pythonModule: PythonModule): PythonOptionsItem =
+        PythonOptionsItem(
+            BspMappings.toBspId(module),
+            emptyList(),
+        )
+
+    override fun dependencySources(targetInfo: TargetInfo, dependencyTree: DependencyTree): Set<URI> =
+        targetInfo.pythonTargetInfo?.run {
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:53]: i think that u r not this `?.run {` here

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        targetInfo.sourcesList.mapNotNull { it.takeIf { it.isExternal } }
+
+    private fun calculateExternalSourcePath(externalSource: FileLocation): URI {
+        val path = bazelPathsResolver.resolveUri(externalSource).toPath()
+        return bazelPathsResolver.resolveUri(findExternalSubdirectory(path) ?: path)
+    }
+
+    private fun findExternalSubdirectory(path: Path?): Path? {
+        return if (path == null || path.parent.endsWith("external"))
+            path
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:71]: if it's null u want to return null or it's a bug?

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    fun toPythonOptionsItem(module: Module, pythonModule: PythonModule): PythonOptionsItem =
+        PythonOptionsItem(
+            BspMappings.toBspId(module),
+            emptyList(),
+        )
+
+    override fun dependencySources(targetInfo: TargetInfo, dependencyTree: DependencyTree): Set<URI> =
+        targetInfo.pythonTargetInfo?.run {
+            dependencyTree.transitiveDependenciesWithoutRootTargets(targetInfo.id)
+                .flatMap(::getExternalSources)
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:55]: i think that u should add filter here to check is the target external

**Review by abrams27 — CHANGES_REQUESTED** [review]

File: bazelrunner/src/main/main2.iml
Comment (abrams27) [inline bazelrunner/src/main/main2.iml:1]: we dont need the `.iml` files - also how did u create them?

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+import kotlin.io.path.toPath
+
+class PythonLanguagePlugin(
+    private val bazelPathsResolver: BazelPathsResolver
+) : LanguagePlugin<PythonModule>() {
+    private var defaultInterpreter: URI? = null
+    private var defaultVersion: String? = null
+
+    override fun prepareSync(targets: Sequence<TargetInfo>) {
+        val targetWithSdk = targets.filter { it.hasPythonTargetInfo() && it.pythonTargetInfo.hasInterpreter() }.firstOrNull()
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:26]:
```suggestion
        val targetWithSdk = targets
          .filter { it.hasPythonTargetInfo() && it.pythonTargetInfo.hasInterpreter() }
          .firstOrNull()
```

Comment (abrams27) [reply server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:26]: and it'd be nice to move the predicate to a separate method

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+        val defaultTargetInfo = targetWithSdk?.pythonTargetInfo
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:27]: i'd just create a method "calculateDetultTargetInfo` and put there this line and the previous one

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+        defaultInterpreter = defaultTargetInfo?.interpreter
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:28]: let's move the calculation to a method

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+            ?.takeUnless { it.relativePath.isNullOrEmpty() }
+            ?.let { bazelPathsResolver.resolveUri(it) }
+        defaultVersion = defaultTargetInfo?.version
+    }
+
+    override fun resolveModule(targetInfo: TargetInfo): PythonModule? =
+        targetInfo.pythonTargetInfo?.run {
+            PythonModule(
+                calculateInterpreterURI(interpreter) ?: defaultInterpreter,
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:37]: let's use named parameters

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+        targetInfo.pythonTargetInfo?.run {
+            dependencyTree.transitiveDependenciesWithoutRootTargets(targetInfo.id)
+                .flatMap(::getExternalSources)
+                .filter(::isExternal)
+                .map(::calculateExternalSourcePath)
+                .toSet()
+        }.orEmpty()
+
+
+    private fun isExternal(fileLocation: FileLocation): Boolean = fileLocation.isExternal
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:69]: i think that u can inline this method

**Review by abrams27 — CHANGES_REQUESTED** [review]
+ old comments

**Review by abrams27 — COMMENTED** [review]
plus all the previous unresolved comments

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt (hunk 12 above)
Comment (mielnikk) [reply server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:85]: we do, cause if the path does not contain site-packages somehow, then the function calling findSitePackagesSubdirectory is going to return the full path

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+        defaultInterpreter = defaultTargetInfo?.interpreter
+            ?.takeUnless { it.relativePath.isNullOrEmpty() }
+            ?.let { bazelPathsResolver.resolveUri(it) }
+        defaultVersion = defaultTargetInfo?.version
+    }
+
+    private fun calculateDefaultTargetInfo(targets: Sequence<TargetInfo>): PythonTargetInfo? {
+        val targetWithSdk = targets.filter(::hasPythonInterpreter).firstOrNull()
+        return targetWithSdk?.pythonTargetInfo
+    }
```
Comment (abrams27) [inline server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:36]:
```suggestion
    private fun calculateDefaultTargetInfo(targets: Sequence<TargetInfo>): PythonTargetInfo? =
        targets
          .filter(::hasPythonInterpreter)
          .firstOrNull()
          ?.pythonTargetInfo
```

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt (hunk 12 above)
Comment (abrams27) [reply server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:85]: okey then let's rewrite it using `when` and make it a bit more explicite + add a tail recursion

Comment (mielnikk) [reply server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt:85]: it is a tail recursion though, it's the equivalent of removing the last directory from the path up to "site-packages"

Comment (abrams27) [reply server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/python/PythonLanguagePlugin.kt]: yes yes, u just need to mark it as tailrec to allow compiler do the magic: https://kotlinlang.org/docs/functions.html#tail-recursive-functions

**Review by abrams27 — APPROVED** [review]

**Comment by abrams27** [comment]
thanks a lot!

**Pull request merged**